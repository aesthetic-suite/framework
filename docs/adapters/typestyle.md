# TypeStyle

Provides [TypeStyle](https://github.com/threepointone/typestyle) support.

```ts
import TypeStyleAesthetic from 'aesthetic-adapter-typestyle';
import { TypeStyle } from 'typestyle';
import { Theme } from './types';

const aesthetic = new TypeStyleAesthetic<Theme>(new TypeStyle({ autoGenerateTag: true }), options);
```

## Installation

```
yarn add aesthetic aesthetic-adapter-typestyle typestyle
// Or
npm install aesthetic aesthetic-adapter-typestyle typestyle
```
