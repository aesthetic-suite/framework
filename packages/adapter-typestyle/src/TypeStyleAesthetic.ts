import Aesthetic, { AestheticOptions, ClassName, Ruleset, Sheet } from 'aesthetic';
import { TypeStyle } from 'typestyle';
import { NativeBlock, ParsedBlock } from './types';

export default class TypeStyleAesthetic<Theme extends object> extends Aesthetic<
  Theme,
  NativeBlock,
  ParsedBlock
> {
  typeStyle: TypeStyle;

  keyframes: { [animationName: string]: ClassName } = {};

  constructor(typeStyle: TypeStyle, options: Partial<AestheticOptions> = {}) {
    super(options);

    this.typeStyle = typeStyle;

    this.syntax
      .on('attribute', this.handleNested)
      .on('css', this.handleCss)
      .on('fallback', this.handleFallback)
      .on('font-face', this.handleFontFace)
      .on('global', this.handleGlobal)
      .on('keyframe', this.handleKeyframe)
      .on('media', this.handleMedia)
      .on('property', this.handleProperty)
      .on('pseudo', this.handleNested)
      .on('selector', this.handleNested)
      .on('support', this.handleSupport);
  }

  protected transformToClassName(styles: ParsedBlock[]): ClassName {
    return this.typeStyle.style(...styles);
  }

  // https://typestyle.github.io/#/raw/-cssraw-
  private handleCss = (css: string) => {
    this.typeStyle.cssRaw(css);
  };

  // https://typestyle.github.io/#/core/concept-fallbacks
  private handleFallback = (
    ruleset: Ruleset<NativeBlock>,
    name: keyof NativeBlock,
    value: any[],
  ) => {
    let fallbacks: any = ruleset.properties[name] || [];

    if (!Array.isArray(fallbacks)) {
      fallbacks = [fallbacks];
    }

    ruleset.addProperty(name, [...value, ...fallbacks]);
  };

  // https://typestyle.github.io/#/raw/fontface
  private handleFontFace = (sheet: Sheet<NativeBlock>, fontFaces: Ruleset<NativeBlock>[]) => {
    fontFaces.forEach(face => {
      this.typeStyle.fontFace(face.toObject() as any);
    });
  };

  // https://typestyle.github.io/#/raw/cssrule
  private handleGlobal = (
    sheet: Sheet<NativeBlock>,
    selector: string,
    ruleset: Ruleset<NativeBlock>,
  ) => {
    this.typeStyle.cssRule(selector, ruleset.toObject());
  };

  // https://typestyle.github.io/#/core/concept-keyframes
  private handleKeyframe = (
    sheet: Sheet<NativeBlock>,
    keyframe: Ruleset<NativeBlock>,
    animationName: string,
  ) => {
    this.keyframes[animationName] = this.typeStyle.keyframes(keyframe.toObject() as any);
  };

  // https://typestyle.github.io/#/core/concept-media-queries
  private handleMedia = (
    ruleset: Ruleset<NativeBlock>,
    query: string,
    value: Ruleset<NativeBlock>,
  ) => {
    this.handleNested(ruleset, `@media ${query}`, value);
  };

  // https://typestyle.github.io/#/core/concept-interpolation
  private handleNested = (
    ruleset: Ruleset<NativeBlock>,
    selector: string,
    value: Ruleset<NativeBlock>,
  ) => {
    const nest = ruleset.nested.$nest || ruleset.createRuleset('$nest');

    nest.addNested(selector.startsWith('@') ? selector : `&${selector}`, value);

    ruleset.addNested('$nest', nest);
  };

  // https://typestyle.github.io/#/core
  private handleProperty = (ruleset: Ruleset<NativeBlock>, name: keyof NativeBlock, value: any) => {
    if (name === 'animationName') {
      ruleset.addProperty(name, this.syntax.injectKeyframes(value, this.keyframes).join(', '));
    } else {
      ruleset.addProperty(name, value);
    }
  };

  // https://typestyle.github.io/#/core/concept-media-queries
  private handleSupport = (
    ruleset: Ruleset<NativeBlock>,
    query: string,
    value: Ruleset<NativeBlock>,
  ) => {
    this.handleNested(ruleset, `@supports ${query}`, value);
  };
}
