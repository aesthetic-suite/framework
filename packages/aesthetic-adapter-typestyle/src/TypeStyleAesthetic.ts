/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Aesthetic, { AestheticOptions, ClassName, Keyframes, Ruleset, Sheet } from 'aesthetic';
import { TypeStyle } from 'typestyle';
import { NativeBlock, ParsedBlock } from './types';

export default class TypeStyleAesthetic<Theme> extends Aesthetic<Theme, NativeBlock, ParsedBlock> {
  typeStyle: TypeStyle;

  keyframes: { [animationName: string]: ClassName } = {};

  constructor(typeStyle?: TypeStyle, options: Partial<AestheticOptions> = {}) {
    super(options);

    this.typeStyle = typeStyle || new TypeStyle({ autoGenerateTag: true });

    this.syntax
      .on('attribute', this.handleNested)
      .on('fallback', this.handleFallback)
      .on('font-face', this.handleFontFace)
      .on('global', this.handleGlobal)
      .on('keyframes', this.handleKeyframes)
      .on('media', this.handleMedia)
      .on('property', this.handleProperty)
      .on('pseudo', this.handleNested)
      .on('selector', this.handleNested)
      .on('supports', this.handleSupports);
  }

  // https://typestyle.github.io/#/core/concept-fallbacks
  handleFallback = (ruleset: Ruleset<NativeBlock>, name: keyof NativeBlock, value: any[]) => {
    let fallbacks: any = ruleset.properties[name] || [];

    if (!Array.isArray(fallbacks)) {
      fallbacks = [fallbacks];
    }

    ruleset.addProperty(name, [...value, ...fallbacks]);
  };

  // https://typestyle.github.io/#/raw/fontface
  handleFontFace = (sheet: Sheet<NativeBlock>, fontFaces: NativeBlock[]) => {
    fontFaces.map(face => this.typeStyle.fontFace(face));
  };

  // https://typestyle.github.io/#/raw/cssrule
  handleGlobal = (sheet: Sheet<NativeBlock>, selector: string, ruleset: Ruleset<NativeBlock>) => {
    this.typeStyle.cssRule(selector, ruleset.toObject());
  };

  // https: //typestyle.github.io/#/core/concept-keyframes
  handleKeyframes = (
    sheet: Sheet<NativeBlock>,
    keyframes: Keyframes<NativeBlock>,
    animationName: string,
  ) => {
    this.keyframes[animationName] = this.typeStyle.keyframes(keyframes);
  };

  // https://typestyle.github.io/#/core/concept-media-queries
  handleMedia = (ruleset: Ruleset<NativeBlock>, query: string, value: Ruleset<NativeBlock>) => {
    this.handleNested(ruleset, `@media ${query}`, value);
  };

  // https://typestyle.github.io/#/core/concept-interpolation
  handleNested = (ruleset: Ruleset<NativeBlock>, selector: string, value: Ruleset<NativeBlock>) => {
    const nest = ruleset.nested['$nest'] || ruleset.createRuleset('$nest');

    nest.addNested(selector.startsWith('@') ? selector : `&${selector}`, value);

    ruleset.addNested('$nest', nest);
  };

  // https://typestyle.github.io/#/core
  handleProperty = (ruleset: Ruleset<NativeBlock>, name: keyof NativeBlock, value: any) => {
    if (name === 'animationName') {
      ruleset.addProperty(name, this.syntax.injectKeyframes(value, this.keyframes).join(', '));
    } else {
      ruleset.addProperty(name, value);
    }
  };

  // https://typestyle.github.io/#/core/concept-media-queries
  handleSupports = (ruleset: Ruleset<NativeBlock>, query: string, value: Ruleset<NativeBlock>) => {
    this.handleNested(ruleset, `@supports ${query}`, value);
  };

  transformToClassName(styles: (NativeBlock | ParsedBlock)[]): ClassName {
    return this.typeStyle.style(...styles);
  }
}
