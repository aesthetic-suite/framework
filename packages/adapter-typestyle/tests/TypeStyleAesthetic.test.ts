/* eslint-disable jest/expect-expect, no-underscore-dangle */

import { TypeStyle } from 'typestyle';
import {
  cleanupStyleElements,
  renderAndExpect,
  DIRECTIONS,
  FONT_ROBOTO_FLAT_SRC,
  FONT_CIRCULAR_MULTIPLE_FLAT_SRC,
  KEYFRAME_FADE,
  SYNTAX_UNIFIED_LOCAL_FULL,
  SYNTAX_ATTRIBUTE,
  SYNTAX_DESCENDANT,
  SYNTAX_FALLBACKS,
  SYNTAX_FONT_FACE,
  SYNTAX_FONT_FACE_MIXED,
  SYNTAX_FONT_FACE_MULTIPLE,
  SYNTAX_GLOBAL,
  SYNTAX_KEYFRAMES,
  SYNTAX_MEDIA_QUERY,
  SYNTAX_MULTI_SELECTOR,
  SYNTAX_PROPERTIES,
  SYNTAX_PSEUDO,
  SYNTAX_SUPPORTS,
  FONT_ROBOTO,
  SYNTAX_MEDIA_QUERY_NESTED,
  SYNTAX_KEYFRAMES_INLINE,
  SYNTAX_FONT_FACES_INLINE,
  SYNTAX_RAW_CSS,
} from 'aesthetic/lib/testUtils';
import TypeStyleAesthetic from '../src/TypeStyleAesthetic';

describe('TypeStyleAesthetic', () => {
  let instance: TypeStyleAesthetic<any>;

  beforeEach(() => {
    instance = new TypeStyleAesthetic(new TypeStyle({ autoGenerateTag: true }));
  });

  afterEach(() => {
    cleanupStyleElements();
  });

  DIRECTIONS.forEach(dir => {
    // eslint-disable-next-line jest/valid-describe
    describe(dir.toUpperCase(), () => {
      it('converts and transforms inline styles', () => {
        expect(
          instance.transformStyles([{ margin: 0 }, { padding: 2 }], { rtl: dir === 'rtl' }),
        ).toBe('f1rvgqmz');
      });

      describe('global sheet', () => {
        it('handles globals', () => {
          renderAndExpect(instance, SYNTAX_GLOBAL, {}, { dir, global: true });
        });

        it('handles @font-face', () => {
          const spy = jest.spyOn(instance.typeStyle, 'fontFace');

          instance.syntax.convertGlobalSheet(SYNTAX_FONT_FACE, { rtl: dir === 'rtl' });

          expect(spy).toHaveBeenCalledWith(FONT_ROBOTO_FLAT_SRC);
          expect(spy).toHaveBeenCalledTimes(1);
        });

        it('handles mixed @font-face', () => {
          const spy = jest.spyOn(instance.typeStyle, 'fontFace');

          instance.syntax.convertGlobalSheet(SYNTAX_FONT_FACE_MIXED, { rtl: dir === 'rtl' });

          expect(spy).toHaveBeenCalledWith(FONT_ROBOTO_FLAT_SRC);
          expect(spy).toHaveBeenCalledWith(FONT_CIRCULAR_MULTIPLE_FLAT_SRC[0]);
          expect(spy).toHaveBeenCalledTimes(5);
        });

        it('handles multiple @font-face', () => {
          const spy = jest.spyOn(instance.typeStyle, 'fontFace');

          instance.syntax.convertGlobalSheet(SYNTAX_FONT_FACE_MULTIPLE, { rtl: dir === 'rtl' });

          expect(spy).toHaveBeenCalledWith(FONT_CIRCULAR_MULTIPLE_FLAT_SRC[0]);
          expect(spy).toHaveBeenCalledTimes(4);
        });

        it('handles @keyframes', () => {
          const spy = jest.spyOn(instance.typeStyle, 'keyframes');

          instance.syntax.convertGlobalSheet(SYNTAX_KEYFRAMES, { rtl: dir === 'rtl' });

          expect(spy).toHaveBeenCalledWith(KEYFRAME_FADE);
        });
      });

      describe('style sheet', () => {
        it('converts unified syntax to native syntax and transforms to a class name', () => {
          instance.typeStyle.fontFace(FONT_ROBOTO as any);

          instance.keyframes.fade = instance.typeStyle.keyframes(KEYFRAME_FADE as any);

          renderAndExpect(
            instance,
            SYNTAX_UNIFIED_LOCAL_FULL,
            {
              button: {
                margin: 0,
                padding: '6px 12px',
                border: '1px solid #2e6da4',
                borderRadius: 4,
                display: 'inline-block',
                cursor: 'pointer',
                fontFamily: 'Roboto',
                fontWeight: 'normal',
                lineHeight: 'normal',
                whiteSpace: 'nowrap',
                textDecoration: 'none',
                textAlign: 'left',
                backgroundColor: '#337ab7',
                verticalAlign: 'middle',
                color: 'rgba(0, 0, 0, 0)',
                animationName: 'f1gwuh0p',
                animationDuration: '.3s',
                $nest: {
                  '&:hover': {
                    backgroundColor: '#286090',
                    borderColor: '#204d74',
                  },
                  '&::before': {
                    content: '"â˜…"',
                    display: 'inline-block',
                    verticalAlign: 'middle',
                    marginRight: 5,
                  },
                  '@media (max-width: 600px)': {
                    padding: '4px 8px',
                  },
                },
              },
            },
            { dir },
          );
        });

        it('handles properties', () => {
          renderAndExpect(instance, SYNTAX_PROPERTIES, SYNTAX_PROPERTIES, { dir });
        });

        it('handles attribute selectors', () => {
          renderAndExpect(
            instance,
            SYNTAX_ATTRIBUTE,
            {
              attr: {
                display: 'block',
                $nest: {
                  '&[disabled]': {
                    opacity: 0.5,
                  },
                },
              },
            },
            { dir },
          );
        });

        it('handles descendant selectors', () => {
          renderAndExpect(
            instance,
            SYNTAX_DESCENDANT,
            {
              list: {
                margin: 0,
                padding: 0,
                $nest: {
                  '&> li': {
                    listStyle: 'bullet',
                  },
                },
              },
            },
            { dir },
          );
        });

        it('handles pseudo selectors', () => {
          renderAndExpect(
            instance,
            SYNTAX_PSEUDO,
            {
              pseudo: {
                position: 'fixed',
                $nest: {
                  '&:hover': {
                    position: 'static',
                  },
                  '&::before': {
                    position: 'absolute',
                  },
                },
              },
            },
            { dir },
          );
        });

        it('handles multiple selectors (comma separated)', () => {
          renderAndExpect(
            instance,
            SYNTAX_MULTI_SELECTOR,
            {
              multi: {
                cursor: 'pointer',
                $nest: {
                  '&:disabled': { cursor: 'default' },
                  '&[disabled]': { cursor: 'default' },
                  '&> span': { cursor: 'default' },
                },
              },
            },
            { dir },
          );
        });

        it('handles inline @keyframes', () => {
          renderAndExpect(
            instance,
            SYNTAX_KEYFRAMES_INLINE,
            dir === 'ltr'
              ? {
                  single: {
                    animationName: 'f1pf291g',
                  },
                  multiple: {
                    animationName: 'f1pf291g, unknown, f1gwuh0p',
                  },
                }
              : {
                  single: {
                    animationName: 'fx4te0v',
                  },
                  multiple: {
                    animationName: 'fx4te0v, unknown, f1gwuh0p',
                  },
                },
            { dir },
          );
        });

        it('handles inline @font-face', () => {
          renderAndExpect(
            instance,
            SYNTAX_FONT_FACES_INLINE,
            {
              single: {
                fontFamily: 'Roboto',
              },
              multiple: {
                fontFamily: 'Circular, OtherFont, Roboto',
              },
            },
            { dir },
          );
        });

        it('handles @fallbacks', () => {
          renderAndExpect(
            instance,
            SYNTAX_FALLBACKS,
            {
              fallback: {
                background: ['red', 'linear-gradient(...)'],
                display: ['block', 'inline-block', 'flex'],
                color: ['blue'],
              },
            },
            { dir },
          );
        });

        it('handles @media', () => {
          renderAndExpect(
            instance,
            SYNTAX_MEDIA_QUERY,
            {
              media: {
                color: 'red',
                paddingLeft: 10,
                $nest: {
                  '@media (max-width: 1000px)': {
                    color: 'green',
                    paddingLeft: 20,
                  },
                  '@media (min-width: 300px)': {
                    color: 'blue',
                    paddingLeft: 15,
                  },
                },
              },
            },
            { dir },
          );
        });

        it('handles nested @media', () => {
          renderAndExpect(
            instance,
            SYNTAX_MEDIA_QUERY_NESTED,
            {
              media: {
                color: 'red',
                $nest: {
                  '@media (min-width: 300px)': {
                    color: 'blue',
                    $nest: {
                      '@media (max-width: 1000px)': {
                        color: 'green',
                      },
                    },
                  },
                },
              },
            },
            { dir },
          );
        });

        it('handles @supports', () => {
          renderAndExpect(
            instance,
            SYNTAX_SUPPORTS,
            {
              sup: {
                display: 'block',
                $nest: {
                  '@supports (display: flex)': {
                    display: 'flex',
                  },
                  '@supports not (display: flex)': {
                    float: 'left',
                  },
                },
              },
            },
            { dir },
          );
        });

        it('handles raw CSS', () => {
          renderAndExpect(instance, SYNTAX_RAW_CSS, {}, { dir });
        });
      });
    });
  });
});
