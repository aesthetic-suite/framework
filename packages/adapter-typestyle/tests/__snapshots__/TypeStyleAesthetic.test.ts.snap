// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeStyleAesthetic global sheet handles globals 1`] = `"body{margin:0}html{height:100%}a{color:red}a:hover{color:darkred}"`;

exports[`TypeStyleAesthetic style sheet converts unified syntax to native syntax and transforms to a class name 1`] = `"@font-face{font-family:Roboto;font-style:normal;font-weight:normal;local:Robo;src-paths:fonts/Roboto.woff2;src-paths:fonts/Roboto.ttf}@keyframes f1gwuh0p{from{opacity:0}to{opacity:1}}.fkp30lv{animation-duration:.3s;animation-name:f1gwuh0p;background-color:#337ab7;border:1px solid #2e6da4;border-radius:4px;color:rgba(0, 0, 0, 0);cursor:pointer;display:inline-block;font-family:Roboto;font-weight:normal;line-height:normal;margin:0;padding:6px 12px;text-align:center;text-decoration:none;vertical-align:middle;white-space:nowrap}.fkp30lv:hover{background-color:#286090;border-color:#204d74}.fkp30lv::before{content:\\"â˜…\\";display:inline-block;margin-right:5px;vertical-align:middle}@media (max-width: 600px){.fkp30lv{padding:4px 8px}}"`;

exports[`TypeStyleAesthetic style sheet handles @fallbacks 1`] = `".f35p7vi{background:red;background:linear-gradient(...);color:blue;display:block;display:inline-block;display:flex}"`;

exports[`TypeStyleAesthetic style sheet handles @media 1`] = `".fuxmg1k{color:red}@media (min-width: 300px){.fuxmg1k{color:blue}}@media (max-width: 1000px){.fuxmg1k{color:green}}"`;

exports[`TypeStyleAesthetic style sheet handles @supports 1`] = `".f6m6wzj{display:block}@supports (display: flex){.f6m6wzj{display:flex}}@supports not (display: flex){.f6m6wzj{float:left}}"`;

exports[`TypeStyleAesthetic style sheet handles attribute selectors 1`] = `".f14zstro{display:block}.f14zstro[disabled]{opacity:0.5}"`;

exports[`TypeStyleAesthetic style sheet handles descendant selectors 1`] = `".f1qve63s{margin:0;padding:0}.f1qve63s> li{list-style:bullet}"`;

exports[`TypeStyleAesthetic style sheet handles multiple selectors (comma separated) 1`] = `".fkkhpct{cursor:pointer}.fkkhpct:disabled,.fkkhpct[disabled],.fkkhpct> span{cursor:default}"`;

exports[`TypeStyleAesthetic style sheet handles nested @media 1`] = `".fuxmg1k{color:red}@media (min-width: 300px){.fuxmg1k{color:blue}@media (max-width: 1000px){.fuxmg1k{color:green}}}"`;

exports[`TypeStyleAesthetic style sheet handles properties 1`] = `".f1tzsa69{color:black;display:inline;margin:10px}"`;

exports[`TypeStyleAesthetic style sheet handles pseudo selectors 1`] = `".fmow1iy{position:fixed}.fmow1iy:hover{position:static}.fmow1iy::before{position:absolute}"`;
